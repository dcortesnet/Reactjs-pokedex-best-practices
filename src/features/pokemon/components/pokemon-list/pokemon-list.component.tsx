import React, { useContext, useEffect, useState } from "react";
import { PokemonContext } from "../../../../shared/context/pokemon.context";
import * as pokemonService from "../../../../shared/services/pokemon.service";
import { PokemonListLayout } from "./pokemon-list.layout";

export const PokemonListComponent = () => {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const { pokemon, setPokemon, setNextPage } = useContext(PokemonContext);

  useEffect(() => {
    getPokemonList();
  }, []);

  const getPokemonList = async () => {
    try {
      setLoading(true);
      const page = await pokemonService.getPokemonPage();
      const pokemon = await pokemonService.getAllPokemonByPage(page);
      setNextPage(page.next);
      setPokemon(pokemon);
      setError(false);
    } catch (error) {
      setError(true);
    } finally {
      setLoading(false);
    }
  };

  return (
    <PokemonListLayout
      loading={loading}
      error={error}
      pokemon={pokemon}
    ></PokemonListLayout>
  );
};
