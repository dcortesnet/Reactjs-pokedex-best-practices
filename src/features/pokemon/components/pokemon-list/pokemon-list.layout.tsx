import { ErrorMessageLayout } from "../../../../shared/components/error-message/error-message.layout";
import { LoadingSpinnerLayout } from "../../../../shared/components/loading-spinner/loading-spinner.layout";
import { Pokemon } from "../../interfaces/pokemon.interface";
import { PokemonCardLayout } from "../pokemon-card/pokemon-card.layout";
import "./pokemon-list.css";

interface Props {
  loading: boolean;
  error: boolean;
  pokemon: Pokemon[];
}

export const PokemonListLayout: React.FC<Props> = ({
  loading,
  error,
  pokemon,
}) => {
  return (
    <div className="pokemon__list">
      {loading ? (
        <LoadingSpinnerLayout />
      ) : error ? (
        <ErrorMessageLayout />
      ) : (
        pokemon.map((pokemon: Pokemon, index: any) => (
          <PokemonCardLayout {...pokemon} key={index}></PokemonCardLayout>
        ))
      )}
    </div>
  );
};
