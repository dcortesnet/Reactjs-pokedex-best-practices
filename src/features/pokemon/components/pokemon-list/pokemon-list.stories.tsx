import { Story, Meta } from "@storybook/react";
import { PokemonListLayout } from "./pokemon-list.layout";
import { pokemonMock } from "../../mocks/pokemon.mock";
export default {
  title: "Pokemon/components/PokemonList",
  component: PokemonListLayout,
} as Meta;

const arrayPok = [pokemonMock, pokemonMock];
export const Template: Story<any> = (args) => (
  <PokemonListLayout pokemon={arrayPok} {...args} />
);
