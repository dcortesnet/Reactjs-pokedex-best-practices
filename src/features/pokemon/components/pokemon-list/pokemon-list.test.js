import { PokemonListLayout } from "./pokemon-list.layout";
import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { pokemonMock } from "../../mocks/pokemon.mock";
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() });

describe("PokemonListLayout", () => {
  it("Renders correctly", () => {
    const r = shallow(
      <PokemonListLayout pokemon={[pokemonMock]}></PokemonListLayout>
    );

    expect(toJson(r)).toMatchSnapshot();
  })
  it("Renders <PokemonCardLayout>", () => {
    const r = shallow(
      <PokemonListLayout pokemon={[pokemonMock]}></PokemonListLayout>
    );

    expect(r.find("PokemonCardLayout").exists()).toBeTruthy();
  });

  it("Renders <LoadingSpinnerLayout>", () => {
    const r = shallow(<PokemonListLayout loading></PokemonListLayout>);

    expect(r.find("LoadingSpinnerLayout").exists()).toBeTruthy();
  });

  it("Renders <ErrorMessageLayout>", () => {
    const r = shallow(<PokemonListLayout error></PokemonListLayout>);

    expect(r.find("ErrorMessageLayout").exists()).toBeTruthy();
  });
});
