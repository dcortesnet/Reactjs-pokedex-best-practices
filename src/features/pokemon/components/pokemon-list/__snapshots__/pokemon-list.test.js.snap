// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonListLayout Renders correctly 1`] = `
<div
  className="pokemon__list"
>
  <PokemonCardLayout
    abilities={Array []}
    base_experience={101}
    forms={Array []}
    game_indices={Array []}
    height={7}
    held_items={Array []}
    id={1}
    is_default={true}
    key="0"
    location_area_encounters="url"
    moves={Array []}
    name="bulbasaur"
    order={1}
    past_types={Array []}
    species="species"
    sprites={
      Object {
        "front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
        "other": Object {
          "dream_world": Object {
            "front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
          },
        },
      }
    }
    stats={Array []}
    types={
      Array [
        Object {
          "slot": 1,
          "type": Object {
            "name": "grass",
            "url": "https://pokeapi.co/api/v2/type/12/",
          },
        },
        Object {
          "slot": 2,
          "type": Object {
            "name": "poison",
            "url": "https://pokeapi.co/api/v2/type/4/",
          },
        },
      ]
    }
    weight={69}
  />
</div>
`;
